components/external_storage/src/s3.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::io::{Error, ErrorKind, Read, Result};
+
+use rusoto_core::region;
+use rusoto_core::request::DispatchSignedRequest;
+use rusoto_core::request::{HttpClient, HttpConfig};
+use rusoto_core::RusotoError;
+use rusoto_credential::StaticProvider;
+use rusoto_s3::*;
+
+use super::ExternalStorage;
+use kvproto::backup::S3 as Config;
+
+/// S3 compatible storage
+#[derive(Clone)]
+pub struct S3Storage {
+    config: Config,
+    client: S3Client,
+}
+
+impl S3Storage {
+    /// Create a new S3 storage for the given config.
+    pub fn new(config: &Config) -> Result<S3Storage> {
+        // This can greatly improve performance dealing with payloads greater
+        // than 100MB. See https://github.com/rusoto/rusoto/pull/1227
+        // for more information.
+        let mut http_config = HttpConfig::new();
+        http_config.read_buf_size(1024 * 1024 * 2);
+        let http_dispatcher = HttpClient::new_with_config(http_config).unwrap();
+
+        S3Storage::with_request_dispatcher(config, http_dispatcher)
+    }
+
+    fn with_request_dispatcher<D>(config: &Config, dispatcher: D) -> Result<S3Storage>
+    where
+        D: DispatchSignedRequest + Send + Sync + 'static,
+        D::Future: Send,
+    {
+        if config.bucket.is_empty() {
+            return Err(Error::new(ErrorKind::InvalidInput, "missing bucket name"));
+        }
+        if config.prefix.contains("/") {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                format!("multi-prefix is not allowed: {}", config.prefix),
+            ));
+        }
```

Comment:
why not?

Reply:
This is from @overvenus's initial commit. I prefer to allow users to decide where to store the backup files.